lib/mayaUsd/utils/converter.h

```diff
@@ -0,0 +1,461 @@
+        using VtValueToMDataHandleFn
+            = std::add_pointer<void(const VtValue&, MDataHandle&, const ConverterArgs&)>::type;
+
+        //! \brief  Construct new converter. This should never be called directly other than during
+        //! generation of all supported converters.
+        Converter(
+            const SdfValueTypeName& usdTypeName,
+            MPlugToUsdAttrFn        plugToAttr,
+            UsdAttrToMPlugFn        attrToPlug,
+            UsdAttrToMDGModifierFn  attrToModifier,
+            MPlugToVtValueFn        plugToVtValue,
+            VtValueToMPlugFn        vtValueToPlug,
+            VtValueToMDGModifierFn  vtValueToModifier,
+            MDataHandleToUsdAttrFn  handleToAttr,
+            UsdAttrToMDataHandleFn  attrToHandle,
+            MDataHandleToVtValueFn  handleToVtValue,
+            VtValueToMDataHandleFn  vtValueoHandle)
+            : _usdTypeName(usdTypeName)
+            , _plugToAttr(plugToAttr)
+            , _attrToPlug(attrToPlug)
+            , _attrToModifier(attrToModifier)
+            , _plugToVtValue { plugToVtValue }
+            , _vtValueToPlug { vtValueToPlug }
+            , _vtValueToModifier(vtValueToModifier)
+            , _handleToAttr(handleToAttr)
+            , _attrToHandle(attrToHandle)
+            , _handleToVtValue { handleToVtValue }
+            , _vtValueoHandle { vtValueoHandle }
+        {
+        }
+
+        //! \brief  Default constructor makes no sense for this class
+        Converter() = delete;
+
+        //! \brief  Nothing special to do in the destructor.
+        ~Converter() = default;
+
+        //! \brief  Look for converter which allows translation for given \p typeName
+        //! \return Valid pointer if conversion is supported.
+        static const Converter* find(const SdfValueTypeName& typeName, bool isArrayPlug);
+        //! \brief  Look for converter which allows translation for given pair of Maya's plug and
+        //! Usd attribute. \return Valid pointer if conversion is supported.
+        static const Converter* find(const MPlug& plug, const UsdAttribute& attr);
+
+        //! \brief  Return sdf value type name this converter is registered for.
+        const SdfValueTypeName& usdType() const { return _usdTypeName; }
+
+        //! \brief  Validate if given pair of Maya's plug and Usd attribute is supported by this
+        //! converter.
+        bool validate(const MPlug& plug, const UsdAttribute& usdAttr) const
+        {
+            return (usdAttr.GetTypeName() == _usdTypeName)
+                && (getUsdTypeName(plug, false) == _usdTypeName);
+        }
+
+        //! \brief  Read current value from given plug and set it on destination attribute. Use
+        //! arguments to control converter behavior, like the time for setting value on attribute.
+        void convert(const MPlug& src, UsdAttribute& dst, const ConverterArgs& args) const
+        {
+            _plugToAttr(src, dst, args);
```

Comment:
Is it possible for a converter not to exist, and that _plugToAttr would still be a nullptr?  Of course in that case you'll get a segmentation fault, but maybe that can never happen, or it's a good enough behavior to indicate a bug?

Reply:
We never generate converters that don't support all these interfaces. My current thinking is that we shouldn't go there and each converter should support them. Time will put this to the test.